pkgs/misc/uboot/default.nix

```diff
@@ -178,6 +179,35 @@ in {
     '';
   };
 
+  ubootNanoPCT4 = buildUBoot rec {
+    # vanilla u-boot does not work with the nanopc-t4
+    # so we use this repo with the required patches applied
+    src = fetchFromGitHub {
+      owner = "tmountain";
+      repo = "u-boot-nanopct4";
+      rev = "23f6f74ec3ba53263ed97ec3cac9979b0ad998bc";
+      sha256 = "07b3gcizkswld796l502bj6ln0hwz7wcm2rp3knpjmmha5llb5dz";
+    };
+
+    version = "2020.10";
+    # this provides wrapped rkbin binaries that are patched to work with NixOS
+    rkbin = callPackage "${src}/rkbin" {};
+
+    defconfig = "nanopc-t4-rk3399_defconfig";
+
+    extraMeta = {
+      platforms = [ "aarch64-linux" ];
+      license = lib.licenses.unfreeRedistributableFirmware;
+    };
+    filesToInstall = [ "idbloader.bin" "uboot.img" "trust.bin" ];
+    postBuild = ''
+      ./tools/mkimage -n rk3399 -T rksd -d ${rkbin}/share/rkbin/rk33/rk3399_ddr_800MHz_v1.24.bin idbloader.bin
+      cat ${rkbin}/share/rkbin/rk33/rk3399_miniloader_v1.19.bin >> idbloader.bin
+      ${rkbin}/bin/trust_merger --replace bl31.elf ${rkbin}/share/rkbin/rk33/rk3399_bl31_v1.30.elf trust.ini
+      ${rkbin}/bin/loaderimage --pack --uboot ./u-boot-dtb.bin uboot.img
```

Comment:
See if the approach used for `ubootRock64` works here.

Really what *may* be needed is the ram training data, everything else is possibly fine with mainline. It may be that using u-boot's SPL works.

So basically replace those three lines with:

```
cat spl/u-boot-spl.bin >> idbloader.img
```

(If it does, you'll need: `filesToInstall = [ "u-boot.itb" "idbloader.img"];`)

Reply:
I tried this suggestion with both the mainline u-boot and with my patched repo, and it did not work. I can submit a new PR as you suggested to move my non-mainline repo to a separate file if you would like.

Reply:
I spoke too soon! The ubootRock64 does work... I was not including the BL31 var.

Reply:
No need for a new PR, push or force-push to this one.

(You might have seen a comment I deleted about a separate file, it was wrong, forget about it...)

Reply:
> I spoke too soon! The ubootRock64 apprach does work... I was not including the BL31 var. I can make a new PR and send it over if you would like. Also, it works with upstream u-boot. No need for the patched repo.

So this looks like it's 100% about RAM training data. Interesting.

Reply:
I made a mistake in my testing. I was calling the idbloader `idbloader.bin` and when I made your suggested changes, the new idbloader was named `idbloader.img`. I was using a small shell script to run the dd commands, so the new `idbloader.img` did not get written to the device / tested when I reported that the `spl/u-boot-spl.bin` was working.

When I tested with the `spl/u-boot-spl.bin` based idbloader, I got nothing (just a black screen).

It took me a while to figure this out, but I caught it when doing a clean ("from scratch") test before submitting my revised PR. The good news is, this PR is still in a much better spot than it was. Here are my findings:

- The ATF bl31.elf works just fine.
- The upstream / unpatched u-boot code works just fine.
- The idbloader.img seems to require the proprietary rockchip blobs.

I'm happy to test further, but my revised PR reflects these changes, and I can confirm that it is working with my NanoPC-T4.

Reply:
Hm, when using `miniloader` from Rockchip, I don't know what actually is going on with regard to the ATF (BL31). Not sure if that miniloader has the ATF, or if it's part of u-boot.itb. My gut feeling is that you can delete the `BL31 = ` line and it will change nothing to the output. That ATF is part of the `u-boot-spl.bin` file.

Reply:
I just tested, and you are correct. I have updated the PR accordingly.